user/api.go

```diff
@@ -714,53 +651,78 @@ func (a *Api) sendError(res http.ResponseWriter, statusCode int, reason string,
 	sendModelAsResWithStatus(res, status.NewStatus(statusCode, reason), statusCode)
 }
 
-func (a *Api) authenticateSessionToken(sessionToken string) (*TokenData, error) {
-	if sessionToken == "" {
+func (a *Api) authenticateToken(r *http.Request) (*TokenData, error) {
+	if sessionToken := r.Header.Get(TP_SESSION_TOKEN); sessionToken != "" {
+		return a.authenticateSessionToken(sessionToken)
+	}
+	return a.accessTokenChecker.Check(r)
+}
+
+func (a *Api) authenticateSessionToken(token string) (*TokenData, error) {
+	if token == "" {
 		return nil, errors.New("Session token is empty")
-	} else if tokenData, err := UnpackSessionTokenAndVerify(sessionToken, a.ApiConfig.Secret); err != nil {
+	}
+	tokenData, err := UnpackSessionTokenAndVerify(token, a.ApiConfig.Secret)
+	if err != nil {
+		if r, rErr := http.NewRequest("GET", "/", nil); rErr == nil {
+			r.Header.Set("authorization", "Bearer "+token)
+			accessTokenData, accessTokenError := a.authenticateAccessToken(r)
+			if accessTokenError == nil {
+				return accessTokenData, nil
+			}
+			log.Println("Error checking as access_token: ", accessTokenError)
+		}
 		return nil, err
-	} else if _, err := a.Store.FindTokenByID(sessionToken); err != nil {
+	}
+	_, err = a.Store.FindTokenByID(token)
+	if err != nil {
 		return nil, err
-	} else {
-		return tokenData, nil
 	}
+	return tokenData, nil
+}
+
+func (a *Api) authenticateAccessToken(r *http.Request) (*TokenData, error) {
+	return a.accessTokenChecker.Check(r)
 }
 
 func (a *Api) tokenUserHasRequestedPermissions(tokenData *TokenData, groupId string, requestedPermissions clients.Permissions) (clients.Permissions, error) {
 	if tokenData.IsServer {
 		return requestedPermissions, nil
 	} else if tokenData.UserId == groupId {
 		return requestedPermissions, nil
-	} else if actualPermissions, err := a.perms.UserInGroup(tokenData.UserId, groupId); err != nil {
```

Comment:
I don't understand the code change for these next few lines. It looks like exactly the same code (just restructured) from before.